### ISSUE TITLE
cc-wrapper mktemp invokation incompatible with MacOS

### ISSUE THREAD

**Issue by aherrmann** [NixOS/nixpkgs#258607]
### Describe the bug
The [`cc-wrapper` invokes `mktemp` with the `--tmpdir` flag](https://github.com/NixOS/nixpkgs/blob/f5892ddac112a1e9b3612c39af1b72987ee5783a/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L249), which [is not supported on MacOS](https://www.unix.com/man-page/osx/1/mktemp/).

### Steps To Reproduce
I do not have a short reproducer. However, the issue was observed on [this GitHub actions CI run](https://github.com/tweag/rules_nixpkgs/actions/runs/6351506068/job/17252837990?pr=423#step:6:1221) in a cgo context.
```
  # runtime/cgo
  mktemp: illegal option -- -
  usage: mktemp [-d] [-q] [-t prefix] [-u] template ...
         mktemp [-d] [-q] [-u] -t prefix 
```

The issue can be resolved by applying the following change to `cc-wrapper.sh`:
```diff
@@ -246,7 +246,7 @@ if [[ -e @out@/nix-support/cc-wrapper-hook ]]; then
 fi

 if (( "${NIX_CC_USE_RESPONSE_FILE:-@use_response_file_by_default@}" >= 1 )); then
-    responseFile=$(mktemp --tmpdir cc-params.XXXXXX)
+    responseFile=$(mktemp "${TMPDIR:-/tmp}/cc-params.XXXXXX")
     trap 'rm -f -- "$responseFile"' EXIT
     printf "%q\n" \
        ${extraBefore+"${extraBefore[@]}"} \
```

### Expected behavior
The `mktemp` error should not have occurred. The code under test used to work, and failed after an update of the nixpkgs revision in use from ea4c80b39be4c09702b0cb3b42eab59e2 to 6500b4580c2a1f3d0f980d32d285739d8e156d92.

### Screenshots
n/a

### Additional context
The commit range includes the following two commits that touch on `mktemp`:
- 520a544ee5f90541b8a457b12d3d17f74f80d515 adds `mktemp --tmpdir` to the strip setup-hook
- 6f2b3ba027f0d74614ba1b21f15ea45b0beb0385 adds `mktemp --tmpdir` to the cc-wrapper for response file handling
The latter one seems to be the cause of the issue.

Similar issues were described in https://github.com/NixOS/nixpkgs/issues/254104#issuecomment-1718173892 and https://github.com/numtide/devshell/issues/10#issuecomment-1516612414.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 21.6.0, macOS 12.6.3`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.13.3`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixpkgs`
```

**Comment by tobim** [comment]
So the wrapper picks up the `mktemp` binary from the system instead of the one from the store because the environment is not pure? This makes the behavior of this and other wrappers unpredictable. Maybe it would be better to make sure that the tools used in the wrapper are actually from its real runtime dependencies in the store?

**Comment by aherrmann** [comment]
@tobim Yes, that matches my understanding.

> Maybe it would be better to make sure that the tools used in the wrapper are actually from its real runtime dependencies in the store?

Yes, that thought crossed my mind, too. However, I saw that there is an [explicit import of coreutils](https://github.com/NixOS/nixpkgs/blob/59998678277cf465e749a144cf707021285bdf82/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L11-L16), which is then [explicitly reverted](https://github.com/NixOS/nixpkgs/blob/59998678277cf465e749a144cf707021285bdf82/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L239-L240). And this import [is conditional](https://github.com/NixOS/nixpkgs/blob/59998678277cf465e749a144cf707021285bdf82/pkgs/build-support/cc-wrapper/default.nix#L90-L91) and seems to be [disabled on MacOS](https://github.com/aherrmann/nixpkgs/blob/59998678277cf465e749a144cf707021285bdf82/pkgs/stdenv/darwin/default.nix#L100). I don't know the context or reasoning for all this logic and didn't venture there. But, that may very well also be a good solution.

**Comment by tobim** [comment]
Ah cool! So it basically already works that way. Setting `nativeTools = false` actually enables the import.

Your explanation led me to find https://github.com/NixOS/nixpkgs/pull/11798 where this logic was first introduced. The discussion there is very interesting, afaiu we need to restore `PATH` to its original value so things like linker scripts will still find the intended executables.

I think that the commit that introduced the `mktemp` should have added it before the path gets restored.

**Comment by aherrmann** [comment]
> Ah cool! So it basically already works that way. Setting `nativeTools = false` actually enables the import.

Oh, I see, thanks! I overlooked one negation.

> I think that the commit that introduced the `mktemp` should have added it before the path gets restored.

That seems like it could solve the issue, too. Perhaps it's easier to just resolve `mktemp` before the `PATH` is reset and then still invoke it later, since [there is some more code in-between](https://github.com/NixOS/nixpkgs/blob/59998678277cf465e749a144cf707021285bdf82/pkgs/build-support/cc-wrapper/cc-wrapper.sh#L239-L249).

I had opened https://github.com/NixOS/nixpkgs/pull/258608 that replaces the `mktemp` invocation by a portable one. That approach was easier to test in my use-case without rebuilding the world.
